prometheus/pkg/apis/sources/v1alpha1/prometheussource_types.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis/duck"
+	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
+	"knative.dev/pkg/kmeta"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PrometheusSource is the Schema for the prometheussources API
+// +k8s:openapi-gen=true
+type PrometheusSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   PrometheusSourceSpec   `json:"spec,omitempty"`
+	Status PrometheusSourceStatus `json:"status,omitempty"`
+}
+
+// Check that Prometheus source can be validated and can be defaulted.
+var _ runtime.Object = (*PrometheusSource)(nil)
+
+// Check that we can create OwnerReferences to a Configuration.
+var _ kmeta.OwnerRefable = (*PrometheusSource)(nil)
+
+// Check that PrometheusSource implements the Conditions duck type.
+var _ = duck.VerifyType(&PrometheusSource{}, &duckv1beta1.Conditions{})
+
+const (
+	// PromQLPrometheusSourceEventType is the PrometheusSource PromQL CloudEvent type.
+	PromQLPrometheusSourceEventType = "dev.knative.prometheus.promql"
+)
+
+// PrometheusSourceSpec defines the desired state of PrometheusSource
+type PrometheusSourceSpec struct {
+	// ServiceAccountName holds the name of the Kubernetes service account
+	// as which the underlying K8s resources should be run. If unspecified
+	// this will default to the "default" service account for the namespace
+	// in which the PrometheusSource exists.
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// ServerURL is the URL of the Prometheus server
+	ServerURL string `json:"serverURL"`
+
+	// PromQL is the Prometheus query for this source
+	PromQL string `json:"promQL"`
+
+	// Sink is a reference to an object that will resolve to a domain
+	// name to use as the sink.
+	// +optional
+	Sink *corev1.ObjectReference `json:"sink,omitempty"`
```

Comment:
This should be a Destination:
https://github.com/knative/eventing/blob/master/docs/spec/sources.md

And we should also have CloudEventOverrides here as well.

Reply:
I'd like to push back on this if that's ok. Can we merge this PR as it stands wrt sinks and ceOverrides and do the conversion of sinks to Destinations across the board in all the sources?
It would be nice to update https://github.com/knative/eventing/blob/master/pkg/duck/sinks.go to retrieve URIs from Destinations according to https://github.com/knative/eventing/blob/master/docs/spec/sources.md, update vendored knative.dev/pkg and knative.dev/eventing in eventing-contrib.

Reply:
I agree w/ @syedriko  on this subject

Reply:
Keeping it consistent and doing conversion and testing (and upgrade docs if needed) for all sources in contrib sounds good to me